pkg/cli/cli.go

```diff
@@ -0,0 +1,271 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cli
+
+import (
+	"fmt"
+	"log"
+	"os"
+
+	"sigs.k8s.io/kubebuilder/pkg/plugin"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold"
+	"sigs.k8s.io/kubebuilder/pkg/scaffold/project"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/pflag"
+)
+
+const (
+	noticeColor = "\033[1;36m%s\033[0m"
+)
+
+// Option is a function that can configure the CLI
+type Option func(*CLI) error
+
+// CLI defines the command line structure and interfaces that are used to
+// scaffold kubebuilder project files.
+type CLI struct {
+	commandName    string
+	defProjVersion string
+	cmd            *cobra.Command
+	extraCommands  []*cobra.Command
+	plugins        map[string]plugin.BasePlugin
+}
+
+// New creates a new CLI instance.
+func New(opts ...Option) (*CLI, error) {
+	c := &CLI{
+		commandName:    "kubebuilder",
+		defProjVersion: project.Version2,
```

Comment:
Given the seeming consensus around name and version, we may need to re-think the `--project-version` flag and this `defProjVersion` field.

If we care about backward compatibility, we could keep and deprecate the `--project-version` flag, and make it automatically choose the "go" plugins. Maybe we don't care since `--project-version` is used during project initialization only (I think?). 

We probably need a new `init` flag, maybe `--plugin`, that takes something like `<name>/<version>`. So for the existing kubebuilder plugins, it would understand `go/v1` and `go/v2`.

Lastly, do we want to have a convention to make it easier to distinguish similar plugins from different authors? Consider two different plugins named `foo`. Should we include a domain component (e.g. `foo.kubebuilder.io` and `foo.operator-framework.io`)? If so, this field starts to look a lot like a Kubernetes group version string (maybe more familiar for users) and it would make it more apparent which binary a given KB project would require. One thing I _don't_ like about this is that it requires more typing from users.

Reply:
@DirectXMan12 @droot thoughts on ^?

Reply:
I kinda imagined it'd go like this:

- `kubebuilder init --project-version v{1,2}` are mostly an alias for `kubebuilder init --plugin go:v{1,2}` (see below)
- `kubebuilder init --plugin NAME:VERSION` (or `NAME/VERSION`) initializes a project with the given plugin, writing that to the project file with the special project version of `3x` or something.  We might want special handling if you actually used `--project-version` to write `1` or `2`
- After `init`, you never specify project version or base plugin again

As for suffixes, might not be a bad idea to encourage it, but I don't expect it'll be a big issue (famous last words) and it is a lot of extra typing.  Maybe allow it as a disambiguation form, but don't require it?

Reply:
The intention was at least that 3x would indicate to aware things that plugins are involved -- if you try to work with a project scaffolded with a particular set of plugins and aren't aware, things could go awry.

We might want to punt on that for later, though.